### ISSUE TITLE
Protect against mock main- and pause-menus

### ISSUE THREAD

**Issue by Desour** [minetest/minetest#12067]
## Problem

In principle, the server has control over what the user experiences. This imposes the problem that it is possible for a server to show a faked main- or pause-menu, and therefore trick the user into sending their password or other sensitive information to the server. To avoid this, there needs to be a way to be sure that one really left the server.

## Solutions

I can see multiple ways to fix this. A combination of all of them could be considered.

### The window title

The window title already says if you are eg. in main-menu or singleplayer. This could be extended to include the pause-menu.

However, the window title can not be seen if you play in fullscreen mode.
And in window mode the title bar can be very small and very far off the center of the screen. All-in-all, it can very easily be overlooked and is therefore not effective for all but the most paranoid users.

### Customisation

One way to prevent the problem is customisation of the main- and pause-menu. If the server doesn't know about the details of this customisation, they can not fake it.

For example, there could be a customizable (via a setting) text somewhere at the top right corner. By default this would be some random hex digits. The user could then set this to some personalized slogan and feel special.
An image would also work.

The problem here is that if the user leaks their customisation (ie. by publishing a screen capture), it looses much of its effect.
Hence, this alone will not be sufficient.

###  A special key

There could be a special key (or key combination) that always brings you back to the pause- or main-menu, and can not be captured by server-controlled content.

An additional benefit of such a key would be the irrevocable ability to escape from the server, even if it always opens a form to capture esc-key presses.

**Comment by sonolosoft** [comment]
If on log out I don't see my installed mods and games, or favourited servers, it would smell phishy. Is this info available to the server?

**Comment by sfan5** [comment]
> There could be a special key (or key combination) that always brings you back to the pause- or main-menu, and can not be captured by server-controlled content.

This should be (and is to some extent) the Escape key.

**Comment by Desour** [comment]
> > There could be a special key (or key combination) that always brings you back to the pause- or main-menu, and can not be captured by server-controlled content.
> 
> This should be (and is to some extent) the Escape key.

The escape key is used to close formspecs if there's one. And mods have reasons to override this behaviour (ie. a confirm close dialog).
The new key must under all circumstances open the pause-menu or go back to main-menu, so this can't really be done by the close-formspec key.
#10380 also needs such a key btw. (ctrl-esc was suggested by sorcerykid).

**Comment by sfan5** [comment]
> The escape key is used to close formspecs if there's one. https://github.com/minetest/minetest/issues/10380 also needs such a key btw.

Yes and if you read my opinion in that issue I am suggesting that Esc should always reach the pause menu, even in the presence of an uncloseable formspec.

**Comment by doxygen-spammer** [comment]
> it is possible for a server to show a faked main- or pause-menu, and therefore trick the user into sending their password or other sensitive information to the server.

Ooof. If I am too silly after playing for hours, I would probably not notice a mocked main menu, even if it lacks certain features.

It would need another security feature, like restricting password inputs to an extra window, which can only be shown before any server is joined.
Then I might learn that such a window should not appear if I loose connection to a server.
The password would then need to be stored in a password manager.
